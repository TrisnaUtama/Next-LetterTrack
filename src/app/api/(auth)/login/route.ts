import { NextResponse, NextRequest } from "next/server";
import prisma from "@/lib/prisma";
import validateUsername from "@/utils/usernameValidate";
import bcrypt from "bcryptjs";
import * as jose from "jose";
import { verifyToken } from "@/lib/validationToken";

export async function GET(request: NextRequest) {
  const employee_id = request.nextUrl.searchParams.get("employee_id");
  const tokenResponse = await verifyToken(request);

  if (tokenResponse instanceof NextResponse) {
    return tokenResponse;
  }

  if (!tokenResponse.success) {
    return NextResponse.json(
      { message: "Unauthorized access" },
      { status: 401 }
    );
  }

  if (!employee_id) {
    return NextResponse.json(
      { message: "Employee ID is required." },
      { status: 400 }
    );
  }
  try {
    const data = await prisma.employee.findUnique({
      where: {
        employee_id: employee_id,
      },
    });
    return NextResponse.json(
      {
        message: "successfully retreived data user login",
        data: data,
      },
      { status: 200 }
    );
  } catch (error: any) {
    console.error("Error during login:", error.message);
    return NextResponse.json(
      { message: "An error occurred while processing your request." },
      { status: 500 }
    );
  }
}

export async function POST(request: NextRequest) {
  const { username, password } = await request.json();

  try {
    if (!validateUsername(username).isValid) {
      return NextResponse.json(
        { message: "Invalid username or password" },
        { status: 400 }
      );
    }

    const user = await prisma.employee.findFirst({
      where: { employee_name: username },
    });

    if (!user) {
      return NextResponse.json(
        { message: "Invalid username or password" },
        { status: 400 }
      );
    }

    const validPassword = await bcrypt.compare(password, user.password);
    if (!validPassword) {
      return NextResponse.json(
        { message: "Invalid username or password" },
        { status: 400 }
      );
    }

    if (!process.env.JWT_SECRET) {
      return NextResponse.json(
        { message: "Server configuration error." },
        { status: 500 }
      );
    }

    const secret = new TextEncoder().encode(process.env.JWT_SECRET);
    const alg = "HS256";
    const jwt = await new jose.SignJWT({ userId: user.employee_id })
      .setProtectedHeader({ alg, typ: "JWT" })
      .setExpirationTime("72h")
      .setSubject(user.employee_id)
      .sign(secret);

    return NextResponse.json({ data: user, token: jwt }, { status: 200 });
  } catch (error: any) {
    console.error("Error during login:", error.message);
    return NextResponse.json(
      { message: "An error occurred while processing your request." },
      { status: 500 }
    );
  }
}
