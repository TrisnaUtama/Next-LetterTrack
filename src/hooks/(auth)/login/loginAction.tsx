"use server";

import { cookies } from "next/headers";

export type LoginResult =
  | { success: true; redirect: string }
  | { success: false; error: string }
  | null;

export default async function loginAction(
  prevState: LoginResult,
  formData: FormData
): Promise<LoginResult> {
  const username = formData.get("username");
  const password = formData.get("password");

  if (!username || !password) {
    return { success: false, error: "Username and password are required" };
  }

  try {
    const res = await fetch(`${process.env.ROOT_URL}/api/login`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ username, password }),
    });

    const data = await res.json();

    if (res.ok) {
      cookies().set("Authorization", data.token, {
        secure: true,
        httpOnly: true,
        expires: new Date(Date.now() + 3 * 24 * 60 * 60 * 1000),
        path: "/",
        sameSite: "strict",
      });

      cookies().set("USER", JSON.stringify(data), {
        secure: true,
        httpOnly: true,
        expires: new Date(Date.now() + 3 * 24 * 60 * 60 * 1000),
        path: "/",
        sameSite: "strict",
      });
      const user_role = data.data.employee_type_id;
      let path;
      if (user_role == 1) {
        path = "/superadmin";
      } else if (user_role == 2) {
        path = "/secretary";
      } else if (user_role == 3) {
        path = "/receptionist";
      } else {
        path = "/basic";
      }
      return { success: true, redirect: path };
    } else {
      return {
        success: false,
        error: data.message || "An error occurred during login",
      };
    }
  } catch (error) {
    console.error("Login error:", error);
    return { success: false, error: "An unexpected error occurred" };
  }
}
