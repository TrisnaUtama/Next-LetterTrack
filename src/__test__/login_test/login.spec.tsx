import "@testing-library/jest-dom";
import Login from "@/app/page";
import LayoutPage from "@/app/layout";
import {
  render,
  screen,
  fireEvent,
  waitFor,
  act,
} from "@testing-library/react";
import { useRouter } from "next/navigation";
import * as loginActionModule from "../../hooks/(auth)/login/loginAction";

jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
}));

jest.mock("../../hooks/(auth)/login/loginAction", () => ({
  __esModule: true,
  default: jest.fn(),
}));

describe("Login Page", () => {
  let push: jest.Mock;

  beforeEach(() => {
    jest.clearAllMocks();

    push = jest.fn();
    (useRouter as jest.Mock).mockReturnValue({ push });
  });

  it("should render the Login page", async () => {
    await act(async () => {
      render(
        <LayoutPage>
          <Login />
        </LayoutPage>
      );
    });

    expect(
      screen.getByRole("heading", { name: /Sign In/i })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("button", { name: /Sign In/i })
    ).toBeInTheDocument();
  });

  it("should submit the form successfully and redirect to the dashboard", async () => {
    (loginActionModule.default as jest.Mock).mockResolvedValueOnce({
      success: true,
      redirect: "/superadmin",
    });

    render(
      <LayoutPage>
        <Login />
      </LayoutPage>
    );

    fireEvent.change(screen.getByLabelText(/username/i), {
      target: { value: "superadmin" },
    });
    fireEvent.change(screen.getByLabelText(/password/i), {
      target: { value: "password" },
    });
    fireEvent.click(screen.getByRole("button", { name: /Sign In/i }));

    await waitFor(() => {
      expect(loginActionModule.default).toHaveBeenCalledWith(
        "superadmin",
        "password"
      );
      expect(push).toHaveBeenCalledWith("/superadmin");
    });
  });

  it("should display an error message when login fails", async () => {
    (loginActionModule.default as jest.Mock).mockResolvedValueOnce({
      success: false,
      error: "Invalid credentials",
    });

    render(
      <LayoutPage>
        <Login />
      </LayoutPage>
    );

    fireEvent.change(screen.getByLabelText(/username/i), {
      target: { value: "wronguser" },
    });
    fireEvent.change(screen.getByLabelText(/password/i), {
      target: { value: "wrongpassword" },
    });
    fireEvent.click(screen.getByRole("button", { name: /Sign In/i }));

    await waitFor(() => {
      expect(screen.getByText("Invalid credentials")).toBeInTheDocument();
    });
  });

  it("matches the snapshot", () => {
    const { asFragment } = render(<Login />);
    expect(asFragment()).toMatchSnapshot();
  });
});
