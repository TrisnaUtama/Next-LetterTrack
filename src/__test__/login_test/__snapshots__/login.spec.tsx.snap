// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page matches the snapshot 1`] = `
<DocumentFragment>
  <main
    class="flex min-h-screen flex-col bg-gradient-to-r from-[#01557B] to-[#019BE1] p-4 md:p-10"
  >
    <div
      class="flex items-center justify-start mb-8"
    >
      <img
        alt="logo"
        data-nimg="1"
        decoding="async"
        height="100"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=256&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=128&q=75 1x, /_next/image?url=%2Fimg.jpg&w=256&q=75 2x"
        style="color: transparent;"
        width="100"
      />
      <p
        class="text-4xl font-bold ml-4"
      >
        <span
          class="text-white"
        >
          Angkasapura |
        </span>
         
        <span
          class="text-[#66B82F]"
        >
          Airports
        </span>
      </p>
    </div>
    <div
      class="flex-grow flex items-center justify-center"
    >
      <div
        class="rounded-xl border bg-card text-card-foreground shadow p-8 w-full max-w-md h-auto"
      >
        <div
          class="flex flex-col space-y-1.5 p-6"
        >
          <h3
            class="font-semibold tracking-tight text-center text-3xl"
          >
            Sign In
          </h3>
          <p
            class="text-muted-foreground text-sm font-semibold text-center"
          >
            More than just tracking, we empower you to
            <br />
            your letters with ease and efficiency
          </p>
        </div>
        <div
          class="p-6 pt-0"
        >
          <form>
            <label
              class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
              for="username"
            >
              Username
            </label>
            <input
              class="flex w-full rounded-md bg-transparent text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 mb-4 mt-1.5 h-10 p-3 font-medium border border-gray-300 focus:border-blue-500 focus:outline-none"
              id="username"
              placeholder="Enter your username"
              required=""
              type="text"
              value=""
            />
            <label
              class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
              for="password"
            >
              Password
            </label>
            <div
              class="relative"
            >
              <input
                class="flex w-full rounded-md bg-transparent text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 mt-1.5 h-10 p-3 font-medium border border-gray-300 focus:border-blue-500 focus:outline-none"
                id="password"
                placeholder="Enter your password"
                required=""
                type="password"
                value=""
              />
              <span
                class="absolute right-3 top-1/2 transform -translate-y-1/2 text-gray-400 cursor-pointer"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-eye "
                  data-icon="eye"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 576 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M288 32c-80.8 0-145.5 36.8-192.6 80.6C48.6 156 17.3 208 2.5 243.7c-3.3 7.9-3.3 16.7 0 24.6C17.3 304 48.6 356 95.4 399.4C142.5 443.2 207.2 480 288 480s145.5-36.8 192.6-80.6c46.8-43.5 78.1-95.4 93-131.1c3.3-7.9 3.3-16.7 0-24.6c-14.9-35.7-46.2-87.7-93-131.1C433.5 68.8 368.8 32 288 32zM144 256a144 144 0 1 1 288 0 144 144 0 1 1 -288 0zm144-64c0 35.3-28.7 64-64 64c-7.1 0-13.9-1.2-20.3-3.3c-5.5-1.8-11.9 1.6-11.7 7.4c.3 6.9 1.3 13.8 3.2 20.7c13.7 51.2 66.4 81.6 117.6 67.9s81.6-66.4 67.9-117.6c-11.1-41.5-47.8-69.4-88.6-71.1c-5.8-.2-9.2 6.1-7.4 11.7c2.1 6.4 3.3 13.2 3.3 20.3z"
                    fill="currentColor"
                  />
                </svg>
              </span>
            </div>
            <div
              class="grid mt-5"
            >
              <button
                class="inline-flex items-center justify-center whitespace-nowrap rounded-md transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 text-primary-foreground shadow h-9 mt-5 p-6 text-lg bg-[#019BE1] hover:bg-[#01557B] font-semibold w-full"
                type="submit"
              >
                Sign In
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </main>
</DocumentFragment>
`;
